Embed static libs

aSemy opened the pull request:
Embed static libs, produced by new workflow (#6 & #7)

Re-enable RocksDbOptions test, and tests on Windows

aSemy committed:
add static libs for each target (so they can be embedded)

aSemy committed:
make custom C functions for AccessHint static

aSemy committed:
update Kotlin/Native cinterop config to use and embed static libs

aSemy committed:
re-enable windows tests

aSemy committed:
re-enable RocksDbOptions tests

aSemy committed:
disable Gradle action setup (libs should be embdedded now, so no need to build them)

aSemy committed:
fix libz/libzlib difference on osx/linix/mingw

aSemy committed:
try adding the linkerOpts L dirs again...

aSemy committed:
update to Gradle 8.0-rc-2

aSemy committed:
try adding more libs

aSemy committed:
maybe fix linking...?

aSemy committed:
disable gradle welcome message

aSemy committed:
include Shlwapi.lib and Rpcrt4.lib in mingw (as detailed in RocksDB wiki)

aSemy committed:
is zlib named zlib or z on Windows...?

aSemy committed:
try adding link libraries on Linux

aSemy committed:
add pkgconfig files

aSemy committed:
try building linux static libs using gcc-9

aSemy committed:
try removing link libs

aSemy committed:
cache more of vcpkg

aSemy committed:
update linux libs with (hopefully) gcc9 versions

aSemy committed:
update linux libs with (hopefully) gcc9 versions

aSemy committed:
only run 'build static libs' if workflow changes

aSemy committed:
try setting gcc 9 on Windows

aSemy committed:
update mingw libs to gcc9 compiled versions (maybe - I have no idea how to confirm)

aSemy committed:
try setting specific gcc minor version

aSemy committed:
try setting gcc 9.2 on linux

aSemy committed:
try building mingw on linux

aSemy committed:
remove install g++9.2

aSemy committed:
disable 'update-alternatives' stuff, disable windows runner

aSemy committed:
try specifying gcc version

aSemy committed:
try to list all gcc versions

aSemy committed:
only install gcc-9

aSemy committed:
revert apt-list > apt list

aSemy committed:
fix gcc version

aSemy committed:
minor formatting improvements in generated code

aSemy committed:
adjust formatting of generated code

aSemy committed:
try manually building rocksdb from source using Makefile...

aSemy committed:
upload logs on failure

aSemy committed:
disable rocksdb-interop-2

aSemy committed:
update vcpkg and rocksdb versions

aSemy committed:
add triplet build logs, and rename to match path

aSemy committed:
only upload logs

aSemy committed:
update rocksdb to 7.9.2 (on linux and macos, still need to figure out how to build for windows...)

aSemy committed:
remove log filter (it seems to filter out everything)

aSemy committed:
exclude empty dirs by default on all copy tasks

aSemy committed:
go back to try building mingw on windows, Linux has some weird error about mutexes

aSemy committed:
try building on linux, and setting alternatives gcc & g++

aSemy committed:
fix alternative for g++

aSemy committed:
try setting up alternative gcc after installing gcc-9

aSemy committed:
more logging for update-alternatives

aSemy committed:
list gcc version

aSemy committed:
rm CC/CXX --version (fails on mac), add which gcc/g++

aSemy committed:
try installing gcc/g++ alternatives...

aSemy committed:
try and build on Windows!

aSemy committed:
update rdb mingw to 7.9.2

aSemy committed:
fix rocksdb version for windows

aSemy committed:
try building static libs on Linux again...

aSemy committed:
try fixing update-alternatives...

aSemy committed:
exclude big 'clean' file from build logs

aSemy committed:
exclude rocksdb/src from build logs

aSemy committed:
exclude all src/.clean from build logs upload

aSemy committed:
try using CC/CXX env vars for building mingw

aSemy committed:
remove -lz link opt

aSemy committed:
exclude exes from build logs

aSemy committed:
exclude binaries from build logs uploader

aSemy committed:
go back to using Windows to build rdb lib...

aSemy committed:
fix cc/cxx on windows

Fuyukai commented:
If you want to use statically linked libraries with K/N you have to build them using K/N's toolchain compiler. For example:

> const: error: undefined reference to 'gettid'

The ``gettid`` wrapper was added to glibc 2.30, but K/N uses glibc 2.15 when linking. You can use the toolchain like so:

```
CC=$(realpath ~/.konan/dependencies/x86_64-unknown-linux-gnu-gcc-8.3.0-glibc-2.19-kernel-4.9-2/bin/x86_64-unknown-linux-gnu-gcc) ./configure
```

aSemy committed:
try using Konan's gcc/g++...

aSemy committed:
use linux to build windows...

aSemy committed:
fix yml error

aSemy committed:
go back to building using Windows!

aSemy committed:
update static libs with libs compiled using konan...

aSemy commented:
@Fuyukai Thanks a lot for the advice! I really appreciate it. 

I tried setting CC and CXX to use the Konan gcc and g++ executables on Windows and Linux, 


https://github.com/aSemy/kotlin-on-the-rocksdb/blob/48f38e046598ff8e1c17083ab00f23b4572529ec/.github/workflows/build_static_libs.yml#L76-L77

https://github.com/aSemy/kotlin-on-the-rocksdb/blob/48f38e046598ff8e1c17083ab00f23b4572529ec/.github/workflows/build_static_libs.yml#L92-L93

and built new static libraries. Unfortunately I see the same errors.

EDIT: I checked the libraries and actually they were built using gcc 11 or something. So I guess vcpkg doesn't acknowledge the CC/CXX variables. 

Also, I didn't setup the Konan download task for the static lib run so I suspect the Konan dir isn't present. What a ball ache this is...

aSemy committed:
try and force cmake to use the konan gcc/g++...

Fuyukai commented:
Yeah you need to download the kotlin-native-prebuilt binaries first. You can just unzip them somewhere and set the compiler path there (I'm not sure where the links are, but the Kotlin gradle plugin prints them on sync if they're not installed). Not sure about vcpkg, I've never used it.

aSemy committed:
try building using Gradle + RocksDB makefile...

aSemy committed:
re-use build_static_libs.yml so GH will actually run the workflow

aSemy committed:
why won't it run?

aSemy committed:
tree .konan

aSemy committed:
fixin'

aSemy committed:
try setting up dummy tasks so Konan downloads dependencies, and try using invariant path separators so it works on bash+windows

aSemy committed:
only tree ./.konan/dependencies

aSemy committed:
try a static lib instead of an exec???

aSemy committed:
manually set Konan data dir

aSemy committed:
ls home dir...

aSemy commented:
I've no idea why but it doesn't look like Kotlin will allow downloading the gcc and g++ on GitHub. When I `ls .konan` the directory that contains gcc and g++ on my local machine doesn't exist

```
$ ls .konan
factory
perflog
runners
warmup
work
kotlin-native-prebuilt-linux-x86_64-1.8.0
$ tree ~/.konan
/home/runner/.konan/dependencies/  [error opening dir]
```

aSemy committed:
try linkNative so Kotlin downloads gcc + g++???

Fuyukai commented:
Here's the direct link: https://download.jetbrains.com/kotlin/native/x86_64-unknown-linux-gnu-gcc-8.3.0-glibc-2.19-kernel-4.9-2.tar.gz

For me Gradle downloads it on the ``compileKotlinLinuxX64`` in one of my projects.

aSemy commented:
thanks! I managed to get the download to trigger, and it built RocksDB for Linux, but it made `librocksdb.a` as 660MB!! That's way too large. the normal size is 30MB.

https://github.com/aSemy/kotlin-on-the-rocksdb/actions/runs/4046467249

aSemy committed:
fix archive name

aSemy committed:
rm pkgconfig dirs

aSemy committed:
try fixing weird errors by making the konan cache workflow specific...

aSemy committed:
maybe the file needs to contain something...?

aSemy committed:
only depend on linkNative task

aSemy committed:
rm unused obj

Fuyukai commented:
>  the normal size is 30MB.

You've built debug libs, which contain a truly incredible amount of debug symbols; just running ``strip`` gets you a 12mb output.

```
l:users@antiskill /t/r/r/lib ./rw
$ ls -lha librocksdb.a
Permissions Size User  Date Modified Name
.rw-r--r--  663M l  1 Feb  1980  librocksdb.a
$ strip librocksdb.a
l:users@antiskill /t/r/r/lib ./rw
$ ls -lha librocksdb.a
Permissions Size User  Date Modified Name
.rw-r--r--   12M l 31 Jan 11:22  librocksdb.a
```

aSemy commented:
Ahhh that makes sense, thanks.

It should be building static libs. The RocksDB docs say to pass in the arg I'm using, `static_lib`, although I do additionally pass in the dependent libraries. 


https://github.com/aSemy/kotlin-on-the-rocksdb/actions/runs/4052346569/jobs/6971673368#step:8:325

```
Starting process 'command 'make''. Working directory: /home/runner/work/kotlin-on-the-rocksdb/kotlin-on-the-rocksdb/externals/lib-rocksdb/build/tmp/rocksDbMakeStaticLib Command: make static_lib libzstd.a libz.a libbz2.a libsnappy.a liblz4.a
```

https://github.com/aSemy/kotlin-on-the-rocksdb/blob/fe1d65fbf2bb8d2ef195a033f684ef4f6b5320de/externals/lib-rocksdb/build.gradle.kts#L156-L166

Fuyukai commented:
They are static libs; they're just unstripped static libs.

aSemy commented:
Oh sorry, I meant that `make static_lib` should build a _release_ static lib

https://github.com/rockset/rocksdb-cloud/blob/master/INSTALL.md#compilation

> [recommended] `make static_lib` will compile librocksdb.a, RocksDB static library. Compiles static library in release mode.

aSemy committed:
remove make -g debug flags

aSemy commented:
The RocksDB Makefile always adds the `-g` flag, which enables the debug symbols, and it's not configurable...

https://github.com/facebook/rocksdb/blob/753d4d507892e8059a17258ea03505dee396b033/Makefile#L307-L308

I've added some custom config to remove the `-g` flags. It seems to work locally...

aSemy committed:
try using a different task to avoid weird mac/windows Kotlin Konan bugs

aSemy committed:
update libs for linux....

aSemy committed:
rm KONAN_DATA_DIR (might be what's screwing up windows and mac compilation?)

aSemy committed:
update mac rocksdb libs

aSemy committed:
rm KONAN_DATA_DIR from Gradle task workflow (seems to be incompatible on Windows...)

aSemy committed:
try disabling Gradle parallel

aSemy commented:
The Linux target finally works 😭 

https://github.com/aSemy/kotlin-on-the-rocksdb/actions/runs/4059380833/jobs/6987361741

Thanks so much for your help @Fuyukai. 

Now, to fix [Windows](https://github.com/aSemy/kotlin-on-the-rocksdb/actions/runs/4059433936/jobs/6987471842#step:8:388)...

```
Starting process 'command 'make''. Working directory: D:\a\kotlin-on-the-rocksdb\kotlin-on-the-rocksdb\externals\lib-rocksdb\build\tmp\rocksDbMakeStaticLib Command: make static_lib rocksdbjavastatic_deps
Successfully started process 'command 'make''
$DEBUG_LEVEL is 0
Unknown platform!
$DEBUG_LEVEL is 0
Unknown platform!
  CC       cache/cache.o
cc1plus.exe: error: unrecognized command-line option '-auxbase-strip'
cc1plus.exe: error: too many filenames given; type 'cc1plus.exe --help' for usage
```

And [macOS](https://github.com/aSemy/kotlin-on-the-rocksdb/actions/runs/4059433939/jobs/6987469173#step:5:90)

```
> command '/Users/runner/work/kotlin-on-the-rocksdb/kotlin-on-the-rocksdb/modules/rocksdb-interop/build/bin/macosX64/debugTest/test.kexe' exited with errors (exit code: 132)
```

aSemy committed:
update macOS libs with those built by the GitHub Workflow

aSemy committed:
update comment for disabling Gradle parallel

aSemy committed:
update comment for disabling Gradle parallel

aSemy committed:
restore vcpkg build for windows

aSemy committed:
fix working dir

aSemy committed:
temp disable the makefile build

aSemy committed:
./gradlew -> gradlew

aSemy committed:
fix workdir

aSemy committed:
fix win setup step

aSemy committed:
rm unnecessary fetch depth (1 is already default)

aSemy committed:
fix triplet setup step

aSemy committed:
use the updated Windows libs (even if strings -a says it's not built with gcc 9.2.0...)

aSemy committed:
use windows paths for gcc/g++

aSemy committed:
use windows paths for gcc/g++

aSemy committed:
use windows paths for gcc/g++

aSemy committed:
use windows paths for gcc/g++

aSemy committed:
escape cmake_c_compiler paths

aSemy committed:
try fixing konan cache for windows

aSemy committed:
use unix path separators in cmake file

aSemy committed:
re-enable vcpkg cache action

aSemy committed:
- try setting up CMAKE_TOOLCHAIN_FILE
- refactor windows + mingw setup

aSemy committed:
update step name

aSemy committed:
use full path for writing konan-toolchain.cmake

aSemy committed:
try fixing location of konan-toolchain.cmake

aSemy committed:
try fixing konan-toolchain.cmake path...

aSemy committed:
fix echo konan-toolchain.cmake

aSemy committed:
move bootstrap step

aSemy committed:
check konan gcc version

aSemy committed:
try using windows path separators for CMAKE_TOOLCHAIN_FILE

aSemy committed:
update path separators, try setting VCPKG_CHAINLOAD_TOOLCHAIN_FILE

aSemy commented:
I've made some progress with Windows compilation using vcpkg (since the Makefile instructions for Windows are complicated, and vcpkg was relatively simply when using the system gcc).

I've told vcpkg to use a custom cmake toolchain file, but it doesn't look like the Konan gcc is compatible with cmake...

```
[1/1] "C:/Program Files/CMake/bin/cmake.exe" -E chdir ".." "C:/Program Files/CMake/bin/cmake.exe" "D:/a/kotlin-on-the-rocksdb/kotlin-on-the-rocksdb/vcpkg/scripts/detect_compiler" "-G" "Ninja" "-DCMAKE_BUILD_TYPE=Release" "-DCMAKE_INSTALL_PREFIX=D:/a/kotlin-on-the-rocksdb/kotlin-on-the-rocksdb/vcpkg/packages/detect_compiler_x64-mingw-static" "-DCMAKE_MAKE_PROGRAM=C:/ProgramData/chocolatey/bin/ninja.exe" "-DCMAKE_SYSTEM_NAME=MinGW" "-DBUILD_SHARED_LIBS=OFF" "-DVCPKG_CHAINLOAD_TOOLCHAIN_FILE=D:/a/kotlin-on-the-rocksdb/kotlin-on-the-rocksdb/konan-toolchain.cmake" "-DVCPKG_TARGET_TRIPLET=x64-mingw-static" "-DVCPKG_SET_CHARSET_FLAG=ON" "-DVCPKG_PLATFORM_TOOLSET=external" "-DCMAKE_EXPORT_NO_PACKAGE_REGISTRY=ON" "-DCMAKE_FIND_PACKAGE_NO_PACKAGE_REGISTRY=ON" "-DCMAKE_FIND_PACKAGE_NO_SYSTEM_PACKAGE_REGISTRY=ON" "-DCMAKE_INSTALL_SYSTEM_RUNTIME_LIBS_SKIP=TRUE" "-DCMAKE_VERBOSE_MAKEFILE=ON" "-DVCPKG_APPLOCAL_DEPS=OFF" "-DCMAKE_TOOLCHAIN_FILE=D:/a/kotlin-on-the-rocksdb/kotlin-on-the-rocksdb/vcpkg/scripts/buildsystems/vcpkg.cmake" "-DCMAKE_ERROR_ON_ABSOLUTE_INSTALL_DESTINATION=ON" "-DVCPKG_CXX_FLAGS=" "-DVCPKG_CXX_FLAGS_RELEASE=" "-DVCPKG_CXX_FLAGS_DEBUG=" "-DVCPKG_C_FLAGS=" "-DVCPKG_C_FLAGS_RELEASE=" "-DVCPKG_C_FLAGS_DEBUG=" "-DVCPKG_CRT_LINKAGE=dynamic" "-DVCPKG_LINKER_FLAGS=" "-DVCPKG_LINKER_FLAGS_RELEASE=" "-DVCPKG_LINKER_FLAGS_DEBUG=" "-DVCPKG_TARGET_ARCHITECTURE=x64" "-DCMAKE_INSTALL_LIBDIR:STRING=lib" "-DCMAKE_INSTALL_BINDIR:STRING=bin" "-D_VCPKG_ROOT_DIR=D:/a/kotlin-on-the-rocksdb/kotlin-on-the-rocksdb/vcpkg" "-DZ_VCPKG_ROOT_DIR=D:/a/kotlin-on-the-rocksdb/kotlin-on-the-rocksdb/vcpkg" "-D_VCPKG_INSTALLED_DIR=D:/a/kotlin-on-the-rocksdb/kotlin-on-the-rocksdb/vcpkg/vcpkg_installed" "-DVCPKG_MANIFEST_INSTALL=OFF"
FAILED: ../CMakeCache.txt 
"C:/Program Files/CMake/bin/cmake.exe" -E chdir ".." "C:/Program Files/CMake/bin/cmake.exe" "D:/a/kotlin-on-the-rocksdb/kotlin-on-the-rocksdb/vcpkg/scripts/detect_compiler" "-G" "Ninja" "-DCMAKE_BUILD_TYPE=Release" "-DCMAKE_INSTALL_PREFIX=D:/a/kotlin-on-the-rocksdb/kotlin-on-the-rocksdb/vcpkg/packages/detect_compiler_x64-mingw-static" "-DCMAKE_MAKE_PROGRAM=C:/ProgramData/chocolatey/bin/ninja.exe" "-DCMAKE_SYSTEM_NAME=MinGW" "-DBUILD_SHARED_LIBS=OFF" "-DVCPKG_CHAINLOAD_TOOLCHAIN_FILE=D:/a/kotlin-on-the-rocksdb/kotlin-on-the-rocksdb/konan-toolchain.cmake" "-DVCPKG_TARGET_TRIPLET=x64-mingw-static" "-DVCPKG_SET_CHARSET_FLAG=ON" "-DVCPKG_PLATFORM_TOOLSET=external" "-DCMAKE_EXPORT_NO_PACKAGE_REGISTRY=ON" "-DCMAKE_FIND_PACKAGE_NO_PACKAGE_REGISTRY=ON" "-DCMAKE_FIND_PACKAGE_NO_SYSTEM_PACKAGE_REGISTRY=ON" "-DCMAKE_INSTALL_SYSTEM_RUNTIME_LIBS_SKIP=TRUE" "-DCMAKE_VERBOSE_MAKEFILE=ON" "-DVCPKG_APPLOCAL_DEPS=OFF" "-DCMAKE_TOOLCHAIN_FILE=D:/a/kotlin-on-the-rocksdb/kotlin-on-the-rocksdb/vcpkg/scripts/buildsystems/vcpkg.cmake" "-DCMAKE_ERROR_ON_ABSOLUTE_INSTALL_DESTINATION=ON" "-DVCPKG_CXX_FLAGS=" "-DVCPKG_CXX_FLAGS_RELEASE=" "-DVCPKG_CXX_FLAGS_DEBUG=" "-DVCPKG_C_FLAGS=" "-DVCPKG_C_FLAGS_RELEASE=" "-DVCPKG_C_FLAGS_DEBUG=" "-DVCPKG_CRT_LINKAGE=dynamic" "-DVCPKG_LINKER_FLAGS=" "-DVCPKG_LINKER_FLAGS_RELEASE=" "-DVCPKG_LINKER_FLAGS_DEBUG=" "-DVCPKG_TARGET_ARCHITECTURE=x64" "-DCMAKE_INSTALL_LIBDIR:STRING=lib" "-DCMAKE_INSTALL_BINDIR:STRING=bin" "-D_VCPKG_ROOT_DIR=D:/a/kotlin-on-the-rocksdb/kotlin-on-the-rocksdb/vcpkg" "-DZ_VCPKG_ROOT_DIR=D:/a/kotlin-on-the-rocksdb/kotlin-on-the-rocksdb/vcpkg" "-D_VCPKG_INSTALLED_DIR=D:/a/kotlin-on-the-rocksdb/kotlin-on-the-rocksdb/vcpkg/vcpkg_installed" "-DVCPKG_MANIFEST_INSTALL=OFF"
System is unknown to cmake, create:
Platform/MinGW to use this system, please post your config file on discourse.cmake.org so it can be added to cmake
-- The C compiler identification is GNU 9.2.0
-- Detecting C compiler ABI info
System is unknown to cmake, create:
Platform/MinGW to use this system, please post your config file on discourse.cmake.org so it can be added to cmake
-- Detecting C compiler ABI info - failed
-- Check for working C compiler: C:/Users/runneradmin/.konan/dependencies/msys2-mingw-w64-x86_64-2/bin/gcc.exe
System is unknown to cmake, create:
Platform/MinGW to use this system, please post your config file on discourse.cmake.org so it can be added to cmake
-- Check for working C compiler: C:/Users/runneradmin/.konan/dependencies/msys2-mingw-w64-x86_64-2/bin/gcc.exe - works
-- Detecting C compile features
-- Detecting C compile features - done
-- The CXX compiler identification is unknown
-- Detecting CXX compiler ABI info
System is unknown to cmake, create:
Platform/MinGW to use this system, please post your config file on discourse.cmake.org so it can be added to cmake
-- Detecting CXX compiler ABI info - failed
-- Check for working CXX compiler: C:/Users/runneradmin/.konan/dependencies/msys2-mingw-w64-x86_64-2/bin/g++.exe
System is unknown to cmake, create:
Platform/MinGW to use this system, please post your config file on discourse.cmake.org so it can be added to cmake
-- Check for working CXX compiler: C:/Users/runneradmin/.konan/dependencies/msys2-mingw-w64-x86_64-2/bin/g++.exe - broken
CMake Error at C:/Program Files/CMake/share/cmake-3.25/Modules/CMakeTestCXXCompiler.cmake:63 (message):
  The C++ compiler

    "C:/Users/runneradmin/.konan/dependencies/msys2-mingw-w64-x86_64-2/bin/g++.exe"

  is not able to compile a simple test program.

  It fails with the following output:

    Change Dir: D:/a/kotlin-on-the-rocksdb/kotlin-on-the-rocksdb/vcpkg/buildtrees/detect_compiler/x64-mingw-static-rel/CMakeFiles/CMakeScratch/TryCompile-0ck1wg
    
    Run Build Command(s):C:/ProgramData/chocolatey/bin/ninja.exe cmTC_8403e && [1/2] Building CXX object CMakeFiles\cmTC_8403e.dir\testCXXCompiler.cxx.obj
    FAILED: CMakeFiles/cmTC_8403e.dir/testCXXCompiler.cxx.obj 
    C:\Users\runneradmin\.konan\dependencies\msys2-mingw-w64-x86_64-2\bin\g++.exe    -o CMakeFiles\cmTC_8403e.dir\testCXXCompiler.cxx.obj -c D:\a\kotlin-on-the-rocksdb\kotlin-on-the-rocksdb\vcpkg\buildtrees\detect_compiler\x64-mingw-static-rel\CMakeFiles\CMakeScratch\TryCompile-0ck1wg\testCXXCompiler.cxx
    cc1plus.exe: error: unrecognized command-line option '-auxbase-strip'
    cc1plus.exe: error: too many filenames given; type 'cc1plus.exe --help' for usage
    ninja: build stopped: subcommand failed.
    
    

  

  CMake will not be able to correctly generate this project.
Call Stack (most recent call first):
  CMakeLists.txt:12 (enable_language)


-- Configuring incomplete, errors occurred!
See also "D:/a/kotlin-on-the-rocksdb/kotlin-on-the-rocksdb/vcpkg/buildtrees/detect_compiler/x64-mingw-static-rel/CMakeFiles/CMakeOutput.log".
See also "D:/a/kotlin-on-the-rocksdb/kotlin-on-the-rocksdb/vcpkg/buildtrees/detect_compiler/x64-mingw-static-rel/CMakeFiles/CMakeError.log".
ninja: build stopped: subcommand failed.
```

aSemy committed:
try enabling verbose cmake

aSemy committed:
try enabling cmake verbose

aSemy committed:
try wrapping the c/cxx compiler paths in quotes

aSemy committed:
update gitignore

aSemy committed:
disable ignored K/N targets warning

aSemy committed:
gradle 8.0.1

aSemy committed:
update git attributes

aSemy committed:
downgrade buildSrc Java to 8 or the project doesn't load

aSemy committed:
jvm 8 fails on GitHub???

aSemy committed:
try using type instead of less de because for whatever reason it is now broken...

aSemy committed:
try fixing the konan-toolchain.cmake file creation (why on earth did this randomly break...?)

aSemy committed:
try using relative path for creating konan-toolchain.cmake

aSemy committed:
replace less with cat

aSemy committed:
update Gradle build config

- Gradle 8.2
- use version catalog
- general tidying

aSemy committed:
add checkWorkflowRocksDBVersion task

aSemy committed:
update vcpkg to 2023.06.20

aSemy committed:
update workflow comments

aSemy committed:
update Kotlin to 1.9.0

aSemy committed:
add ExperimentalForeignApi opt-in

aSemy committed:
use libs.versions.toml

aSemy committed:
tidy RocksDbIterator

aSemy committed:
replace deprecated `getTimeMillis()` in `testUtils.kt#tempDir()`

aSemy committed:
update RocksDbTest.kt, use `runTest {}` util

aSemy committed:
tidy up `rdbWrapperGenerator.kt`, de-duplicate Comparator/ComparatorWithTs

aSemy committed:
rename RdbWrapperGenerator

aSemy committed:
full checkout of vcpkg

aSemy committed:
add tree debug print

aSemy committed:
rm CMAKE_C_COMPILER from konan-toolchain.cmake

try to prevent error `/c/Users/runneradmin/konan-toolchain.cmake: line 1: syntax error near unexpected token `CMAKE_C_COMPILER'`

aSemy committed:
remove 'tree', which isn't available on Windows...

aSemy committed:
remove tree, which isn't available on Windows

aSemy committed:
more logging, try setting CMAKE_C_COMPILER without quotes...

aSemy committed:
tweak logging

aSemy committed:
try fixing VCPKG_CHAINLOAD_TOOLCHAIN_FILE

aSemy committed:
use `$HOME` in VCPKG_CHAINLOAD_TOOLCHAIN_FILE

aSemy committed:
use explicit path instead of `$HOME` for `konan-toolchain.cmake`

aSemy committed:
vcpkg has a weird error (`unrecognized command-line option '-auxbase-strip'`), so try using the Makefile...

aSemy committed:
move `@KotlinOnTheRocksDbInternalApi` to a different file

aSemy committed:
update rdb to 8.3.2

aSemy committed:
implement new `AutoCloseable` in rdb classes

aSemy committed:
disable mingwX64, because of error: `unrecognized command-line option '-auxbase-strip'`

aSemy committed:
tidy up build config, and wrapper generator tasks

aSemy committed:
re-enable parallel (seems to work okay in Kotlin 1.9)

aSemy committed:
disable mingw test workflow

aSemy committed:
tidy build config, and external binaries/headers

aSemy committed:
update language stats exclude

aSemy committed:
disable caches to try fixing tests on GitHub CI/CD

aSemy committed:
disable building build_static_libs.yml, unless the workflow changes

aSemy committed:
rocksdb.def: add libs to compilerOpts, and tidy up

aSemy committed:
mucking about with trying to get the libraries working. Seems to work for Mac and Linux?

aSemy committed:
add `--info` to workflow tests
